auth/auth.go

```diff
@@ -412,115 +411,100 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
 	return user
 }
 
-// Authenticates a user based on a JWT token string and a set of providers.  Attempts to match the
-// issuer in the token with a provider.
-// Used to authenticate a JWT token coming from an insecure source (e.g. client request)
-// If the token is validated but the user for the username defined in the subject claim doesn't exist,
-// creates the user when autoRegister=true.
-func (auth *Authenticator) AuthenticateUntrustedJWT(token string, providers OIDCProviderMap, callbackURLFunc OIDCCallbackURLFunc) (User, jose.JWT, error) {
-
-	base.Debugf(base.KeyAuth, "AuthenticateJWT called with token: %s", base.UD(token))
+// Authenticates a user based on a JWT token string and a set of providers. Attempts to match the issuer in the
+// token with a provider. Used to authenticate a JWT token coming from an insecure source (e.g. client request).
+// If the token is validated but the user for the username defined in the subject claim doesn't exist, creates
+// the user when autoRegister=true.
+func (auth *Authenticator) AuthenticateUntrustedJWT(token string, providers OIDCProviderMap, callbackURLFunc OIDCCallbackURLFunc) (User, *oidc.IDToken, error) {
+	base.Debugf(base.KeyAuth, "AuthenticateUntrustedJWT called with token: %s", base.UD(token))
 
-	// Parse JWT (needed to determine issuer/provider)
-	jwt, err := jose.ParseJWT(token)
+	// Get the ID Token from the raw JWT string; determine issuer/provider.
+	idToken, err := GetIDToken(token)
 	if err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing JWT in AuthenticateJWT: %v", err)
-		return nil, jose.JWT{}, err
-	}
-
-	// Get client for issuer
-	issuer, audiences, err := GetJWTIssuer(jwt)
-	base.Debugf(base.KeyAuth, "JWT issuer: %v, audiences: %v", base.UD(issuer), base.UD(audiences))
-	if err != nil {
-		base.Debugf(base.KeyAuth, "Error getting JWT issuer: %v", err)
-		return nil, jose.JWT{}, err
+		base.Debugf(base.KeyAuth, "Error parsing JWT string to IDToken in AuthenticateUntrustedJWT: %v", err)
+		return nil, &oidc.IDToken{}, err
 	}
 
 	base.Debugf(base.KeyAuth, "Call GetProviderForIssuer w/ providers: %+v", base.UD(providers))
-	provider := providers.GetProviderForIssuer(issuer, audiences)
+	provider := providers.GetProviderForIssuer(idToken.Issuer, idToken.Audience)
```

Comment:
In the case that there's only a single provider defined in providers, the work to extract the issuer and audience from the token manually isn't necessary - we could jump straight to attempted verification using the provider, and let verification fail on the issuer/audience mismatch.  We should refactor to make this the default path (since multiple providers is the uncommon use case), and only do the work to manually extract issuer/audience in the multi-provider case.